### PR TITLE
Charm factorial

### PR DESCRIPTION
**Pull request by MyatCharm**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations
- [ ] The function is not called in the function file

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by PyaeLinn01 — APPROVED** [review]
Well done.

File: solutions/factorial.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+def factorial(n):
```
Comment (PyaeLinn01) [inline solutions/factorial.py:1]: The docstring is clear and well-structured, including the purpose, arguments, return value, and exceptions.
Good use of the @author and Created on tags for documentation.

File: solutions/tests/test_factorial.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+from ..factorial import factorial
+
+
+class TestFactorial(unittest.TestCase):
+    """Unit tests for the factorial function."""
+
+    def test_factorial_zero(self):
+        """Test factorial of 0."""
+        self.assertEqual(factorial(0), 1)
+
```
Comment (PyaeLinn01) [inline solutions/tests/test_factorial.py:25]: Add a test to ensure that the function raises a ValueError for negative numbers.

def test_factorial_negative(self):
    """Test factorial with a negative number."""
    with self.assertRaises(ValueError):
        factorial(-1)

File: solutions/tests/test_factorial.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    def test_factorial_positive(self):
+        """Test factorial of a positive number."""
+        self.assertEqual(factorial(5), 120)
+        self.assertEqual(factorial(3), 6)
+        self.assertEqual(factorial(1), 1)
+
+    def test_factorial_large_number(self):
+        """Test factorial of a larger number."""
+        self.assertEqual(factorial(10), 3628800)
+
```
Comment (PyaeLinn01) [inline solutions/tests/test_factorial.py:35]: Add tests to handle invalid types (e.g., floats, strings, lists). This ensures the function raises a TypeError when appropriate.

def test_factorial_non_integer(self):
    """Test factorial with non-integer input."""
    with self.assertRaises(TypeError):
        factorial(3.5)
    with self.assertRaises(TypeError):
        factorial("string")
    with self.assertRaises(TypeError):
        factorial([1, 2, 3])

Comment (PyaeLinn01) [inline solutions/tests/test_factorial.py:35]: Add a test for a very large input to verify if the function can handle such cases or throws a stack overflow error for recursive implementations.
